h/services/group.py

```diff
@@ -119,6 +122,57 @@ def groupids_created_by(self, user):
             g.pubid for g in self.session.query(Group.pubid).filter_by(creator=user)
         ]
 
+    @classmethod
+    def _group_filter(cls, query, pubids_or_groupids: List[str]):
+        """
+        Apply a filter to a query to efficiently limit it to certain groups.
+
+        This works for a mixed list of pubids or groupids.
+        """
+
+        # We will build up filter clauses for a single efficient query
+        clauses = []
+
+        if pubids := [
+            id_ for id_ in pubids_or_groupids if not group_util.is_groupid(id_)
+        ]:
+            clauses.append(cls._in_or_equals(Group.pubid, pubids))
+
+        if groupids := [
+            id_ for id_ in pubids_or_groupids if group_util.is_groupid(id_)
+        ]:
+            by_authority = defaultdict(list)
+            for parts in (group_util.split_groupid(groupid) for groupid in groupids):
+                by_authority[parts["authority"]].append(parts["authority_provided_id"])
+
+            # We'll probably only ever get called with a single authority, but
+            # there's no harm in being able to support multiple
+            for authority, authority_provided_ids in by_authority.items():
+                clauses.append(
+                    sa.and_(
+                        Group.authority == authority,
+                        cls._in_or_equals(
+                            Group.authority_provided_id, authority_provided_ids
+                        ),
+                    )
+                )
+
+        if len(clauses) == 1:
```

Comment:
Looks like SQLA is already doing this itself (example from LMS)

```
print(db.query(models.LTIRegistration).filter(sa.or_(models.LTIRegistration.client_id=="")))
SELECT lti_registration.created AS lti_registration_created, lti_registration.updated AS lti_registration_updated, lti_registration.id AS lti_registration_id, lti_registration.issuer AS lti_registration_issuer, lti_registration.client_id AS lti_registration_client_id, lti_registration.auth_login_url AS lti_registration_auth_login_url, lti_registration.key_set_url AS lti_registration_key_set_url, lti_registration.token_url AS lti_registration_token_url 
FROM lti_registration 
WHERE lti_registration.client_id = %(client_id_1)s

In [14]: print(db.query(models.LTIRegistration).filter(sa.or_(models.LTIRegistration.client_id=="", models.LTIRegistration.issuer=="")))
SELECT lti_registration.created AS lti_registration_created, lti_registration.updated AS lti_registration_updated, lti_registration.id AS lti_registration_id, lti_registration.issuer AS lti_registration_issuer, lti_registration.client_id AS lti_registration_client_id, lti_registration.auth_login_url AS lti_registration_auth_login_url, lti_registration.key_set_url AS lti_registration_key_set_url, lti_registration.token_url AS lti_registration_token_url 
FROM lti_registration 
WHERE lti_registration.client_id = %(client_id_1)s OR lti_registration.issuer = %(issuer_1)s
```

Reply:
Ok, I'll try and simplify this then.

Reply:
I've done this and it is tidier.